clients/spark/core/src/main/scala/io/treeverse/clients/ApiClient.scala

```diff
@@ -198,28 +219,48 @@ class ApiClient private (conf: APIConfigurations) {
    *  translate that URL to use an appropriate Hadoop FileSystem.
    */
   def getMetaRangeURL(repoName: String, commitID: String): String = {
-    val commit = commitsApi.getCommit(repoName, commitID)
+    val commit = getCommit(repoName, commitID)
     getMetaRangeURL(repoName, commit)
   }
 
   /** Query lakeFS for a URL to the range of rangeID of repoName and
    *  translate that URL to use an appropriate Hadoop FileSystem.
    */
   def getRangeURL(repoName: String, rangeID: String): String = {
-    val range = metadataApi.getRange(repoName, rangeID)
+    val getRange = new dev.failsafe.function.CheckedSupplier[StorageURI]() {
+      def get(): StorageURI = metadataApi.getRange(repoName, rangeID)
+    }
+    val range = retryWrapper.wrapWithRetry(getRange)
     val location = range.getLocation
     URI
       .create(getStorageNamespace(repoName, StorageClientType.HadoopFS) + "/" + location)
       .normalize()
       .toString
   }
 
-  def getBranchHEADCommit(repoName: String, branch: String): String =
-    branchesApi.getBranch(repoName, branch).getCommitId
+  def getBranchHEADCommit(repoName: String, branch: String): String = {
+    val getBranch = new dev.failsafe.function.CheckedSupplier[Ref]() {
+      def get(): Ref = branchesApi.getBranch(repoName, branch)
+    }
+    val response = retryWrapper.wrapWithRetry(getBranch)
+    response.getCommitId
+  }
 
   // Instances of case classes are compared by structure and not by reference https://docs.scala-lang.org/tour/case-classes.html.
   case class StorageNamespaceCacheKey(
       repoName: String,
       storageClientType: StorageClientType
   )
 }
+
+class RequestRetryWrapper(val numRetries: Int = 5) {
+    private val retryPolicy : Policy[Any] = RetryPolicy.builder()
+      .withDelay(Duration.ofMillis(500))
+      .withMaxRetries(numRetries)
+      .build()
+
+  def wrapWithRetry[T](fn: CheckedSupplier[T]): T = {
+      Failsafe.`with`(Seq(retryPolicy).asJava).get(fn)
```

Comment:
This creates a new `FailsafeExecutor` per request. Not sure it's needed. You can create a `FailsafeExecutor[Any]` once (for example, in the ApiClient's companion object).

Then you just call it like:
```
    val repo = executor.get(() => repositoriesApi.getRepository(key.repoName))
```